lib/connectors/memory.js

```diff
@@ -351,28 +351,46 @@ Memory.prototype.fromDb = function(model, data) {
   data = deserialize(data);
   var props = this._models[model].properties;
   for (var key in data) {
-    var val = data[key];
-    if (val === undefined || val === null) {
-      continue;
-    }
-    if (props[key]) {
-      switch (props[key].type.name) {
-        case 'Date':
-          val = new Date(val.toString().replace(/GMT.*$/, 'GMT'));
-          break;
-        case 'Boolean':
-          val = Boolean(val);
-          break;
-        case 'Number':
-          val = Number(val);
-          break;
-      }
-    }
-    data[key] = val;
+    data[key] = this._castPropertyValue(key, data[key], props);
   }
   return data;
 };
 
+Memory.prototype._castPropertyValue = function(prop, val, props) {
+  var self = this;
+  if (val === undefined || val === null || !props[prop]) {
+    return val;
+  }
+
+  if (Array.isArray(val)) {
+    return val.map(function(val) {
+      return self._castPropertyValue(prop, val, props);
+    });
+  }
+
+  var isArray = Array.isArray(props[prop].type);
+  var propType = isArray ? props[prop].type[0] : props[prop].type;
+
+  switch (propType.name) {
+    case 'Date':
+      val = new Date(val.toString().replace(/GMT.*$/, 'GMT'));
+      break;
+    case 'Boolean':
+      val = Boolean(val);
+      break;
+    case 'Number':
+      val = Number(val);
+      break;
+    case 'ModelConstructor':
+      for (var subProp in val) {
+        val[subProp] = this._castPropertyValue(subProp, val[subProp], propType.definition.properties);
```

Comment:
You're confusing the `type.name` field with the `type.modelName` field. The `type.name` will always be `ModelConstructor`. This is evident in the following screenshot froma debug session in my IDE:
![type.name screenshot](https://user-images.githubusercontent.com/3804041/28136740-af2cc5e6-66ff-11e7-8efa-48dbbcf2bfaf.png)

`propType.definition` and `definition.properties` has been set on every model I've tested.